org.jacoco.core/src/org/jacoco/core/runtime/ClassInjectionRuntime.java

```diff
@@ -0,0 +1,308 @@
+
+	private static final String FIELD_NAME = "data";
+
+	private static final String FIELD_TYPE = "Ljava/lang/Object;";
+
+	private final Class<?> locator;
+
+	private final String className;
+
+	/**
+	 * Creates new instance of runtime.
+	 *
+	 * @param instrumentation
+	 *            instrumentation interface
+	 * @return new runtime instance or <code>null</code> if not Java 9
+	 */
+	public static IRuntime create(final Instrumentation instrumentation) {
+		try {
+			Class.forName("java.lang.Module");
+		} catch (final ClassNotFoundException e) {
+			return null;
+		}
+		try {
+			redefineJavaBaseModule(instrumentation);
+		} catch (final Exception e) {
+			throw new RuntimeException(e);
+		}
+		return new ClassInjectionRuntime(Object.class, "$JaCoCo");
+	}
+
+	/**
+	 * Creates a new runtime which will define a class to the same class loader
+	 * and in the same package and protection domain as given class.
+	 *
+	 * @param locator
+	 *            class to identify the target class loader and package
+	 * @param simpleClassName
+	 *            simple name of the class to be defined
+	 */
+	ClassInjectionRuntime(final Class<?> locator,
+			final String simpleClassName) {
+		this.locator = locator;
+		this.className = locator.getPackage().getName().replace('.', '/') + '/'
+				+ simpleClassName;
+	}
+
+	@Override
+	public void startup(final RuntimeData data) throws Exception {
+		super.startup(data);
+		final Class<?> cls = Lookup //
+				.privateLookupIn(locator, Lookup.lookup()) //
+				.defineClass(createClass(className));
+		Lookup.lookup().findStaticSetter(cls, FIELD_NAME, Object.class)
+				.invokeWithArguments(data);
+	}
+
+	public void shutdown() {
+		// nothing to do
+	}
+
+	public int generateDataAccessor(final long classid, final String classname,
+			final int probecount, final MethodVisitor mv) {
+		mv.visitFieldInsn(Opcodes.GETSTATIC, className, FIELD_NAME, FIELD_TYPE);
+
+		RuntimeData.generateAccessCall(classid, classname, probecount, mv);
+
+		return 6;
+	}
+
+	private static byte[] createClass(final String name) {
+		final ClassWriter cw = new ClassWriter(0);
+		cw.visit(Opcodes.V9, Opcodes.ACC_SYNTHETIC | Opcodes.ACC_PUBLIC,
+				name.replace('.', '/'), null, "java/lang/Object", null);
+		cw.visitField(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC, FIELD_NAME,
+				FIELD_TYPE, null, null);
+		cw.visitEnd();
+		return cw.toByteArray();
+	}
+
+	/**
+	 * <pre>
+	 * {@code
+	 * instrumentation.redefineModule(
+	 *   getModule(Object.class),
+	 *   Collections.emptySet(),
+	 *   Collections.emptyMap(),
+	 *   Collections.singletonMap(
+	 *     "java.lang",
+	 *     Collections.singleton(getModule(ClassInjectionRuntime.class))),
+	 *   Collections.emptySet(),
+	 *   Collections.emptyMap())
+	 * }
+	 * </pre>
+	 */
+	private static void redefineJavaBaseModule(
+			final Instrumentation instrumentation) throws Exception {
+		Lookup.lookup() //
+				.bind( //
+						instrumentation, //
+						"redefineModule", //
```

Comment:
@marchof doh, not only reflection, but even this doesn't work in case of `--illegal-access=deny` 😢

Reply:
In fact

```
instrumentation.getClass().getMethod(...).invoke(...)
```

and

```
Lookup.lookup().bind(instrumentation, ...).invokeWithArguments(...)
```

are illegal, whereas

```
java.lang.Instrumentation.class.getMethod(...).invoke(...)
```

and

```
Lookup.lookup().findVirtual(java.lang.Instrumentation.class, ...).invokeWithArguments(...)
```

are legal, because `instrumentation.getClass()` is `sun.instrument.InstrumentationImpl` and

```
module java.instrument {
    exports java.lang.instrument;

    // allow java launcher to load agents in executable JAR files
    exports sun.instrument to java.base;
}
```